iaso/models/tenant_users.py

```diff
@@ -1,5 +1,64 @@
 from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
 from django.db import models
+from django.db.models import Q
+
+from iaso.models import Account
+
+
+def slugify_account(name: str) -> str:
+    return name.lower().replace(" ", "_")
+
+
+class TenantUserManager(models.Manager):
+    def create_user_or_tenant_user(
+        self, username: str, email: str, first_name: str, last_name: str, account: Account
+    ) -> User:
+        existing_user = User.objects.filter(username=username).first()
+
+        if not existing_user:
+            user = User.objects.create(
+                username=username,
+                email=email,
+                first_name=first_name,
+                last_name=last_name,
+            )
+            return user
+
+        if hasattr(existing_user, "iaso_profile") and existing_user.iaso_profile.account == account:
+            raise ValidationError("Username already exists for this account.")
+
+        existing_tenant_user = self.filter(Q(main_user=existing_user) | Q(account_user=existing_user)).first()
+
+        if existing_tenant_user:
+            main_user = existing_tenant_user.main_user
+
+        elif hasattr(existing_user, "iaso_profile"):  # Insert `existing_user` and its profile into `TenantUser`.
+            existing_account = existing_user.iaso_profile.account
+            new_username = f"{username}_{slugify_account(existing_account.name)}"
+            existing_user.username = new_username
+            existing_user.save()
+            main_user = User.objects.create(
+                username=username,
+                email=existing_user.email,
+                first_name=existing_user.first_name,
+                last_name=existing_user.last_name,
+            )
+            self.create(main_user=main_user, account_user=existing_user)
+
+        else:
+            main_user = existing_user
+
+        new_username = f"{username}_{slugify_account(account.name)}"
+        user = User.objects.create(
+            username=new_username,
+            email=email,
+            first_name=first_name,
+            last_name=last_name,
+        )
+        self.create(main_user=main_user, account_user=user)
+
+        return user
```

Comment:
@bramj WDYT of this logic?

My goal would be to use this code everywhere and replace the logic here (https://github.com/BLSQ/iaso/blob/48da9a9f1093a6c4dc74d5c8e6798bf611c3b4c4/iaso/api/profiles/profiles.py#L355-L392) by this function.

Reply:
It makes my brain hurt but it looks correct to me!  :+1: